internal/credit/postgres_connector/grant.go

```diff
@@ -0,0 +1,401 @@
+		// use the last reset as the high watermark
+		q = q.Where(func(s *sql.Selector) {
+			s.LeftJoin(subQuery).
+				On(s.C(db_credit.FieldNamespace), t.C(db_credit.FieldNamespace)).
+				On(s.C(db_credit.FieldSubject), t.C(db_credit.FieldSubject))
+
+			// Ensure the effective date is greater than the last reset date
+			s.Where(
+				sql.Or(
+					sql.IsNull(subQuery.C("highwatermark")),
+					sql.ColumnsGTE(t.C(db_credit.FieldEffectiveAt), subQuery.C("highwatermark")),
+				),
+			)
+		})
+	}
+	if params.IncludeVoid {
+		// Has no void children or is void
+		q = q.Where(
+			db_credit.Or(
+				db_credit.EntryTypeEQ(credit_model.EntryTypeVoidGrant),
+				db_credit.And(
+					db_credit.EntryTypeEQ(credit_model.EntryTypeGrant),
+					db_credit.Not(db_credit.HasChildrenWith(
+						db_credit.EntryTypeEQ(credit_model.EntryTypeVoidGrant),
+					)),
+				),
+			),
+		)
+	} else {
+		// Has no void children
+		q = q.Where(
+			db_credit.EntryTypeEQ(credit_model.EntryTypeGrant),
+			db_credit.Not(db_credit.HasChildrenWith(
+				db_credit.EntryTypeEQ(credit_model.EntryTypeVoidGrant),
+			)),
+		)
+	}
+	entities, err := q.All(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list grants: %w", err)
+	}
+
+	var list []credit_model.Grant
+	for _, entity := range entities {
+		grant, err := mapGrantEntity(entity)
+		if err != nil {
+			return nil, fmt.Errorf("failed to map grant entity: %w", err)
+		}
+		list = append(list, grant)
+	}
+
+	return list, nil
+}
+
+func (c *PostgresConnector) GetGrant(ctx context.Context, namespace string, id string) (credit_model.Grant, error) {
+	entity, err := c.db.CreditEntry.Query().Where(
+		db_credit.Or(
+			// grant
+			db_credit.And(
+				db_credit.Namespace(namespace),
+				db_credit.ID(id),
+				db_credit.EntryTypeEQ(credit_model.EntryTypeGrant),
+				db_credit.Not(db_credit.HasChildren()),
+			),
+			// void grant
+			db_credit.And(
+				db_credit.Namespace(namespace),
+				db_credit.HasParentWith(db_credit.ID(id)),
+				db_credit.EntryTypeEQ(credit_model.EntryTypeVoidGrant),
+			),
+		),
+	).Only(ctx)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return credit_model.Grant{}, &credit_model.GrantNotFoundError{GrantID: id}
+		}
+
+		return credit_model.Grant{}, fmt.Errorf("failed to get grant: %w", err)
+	}
+
+	grantOut, err := mapGrantEntity(entity)
+	if err != nil {
+		return credit_model.Grant{}, fmt.Errorf("failed to map grant entity: %w", err)
+	}
+	return grantOut, nil
+}
+
+// Reset resets the ledger for the subject.
+// Rolls over grants with rollover configuration.
+func (c *PostgresConnector) Reset(ctx context.Context, namespace string, reset credit_model.Reset) (credit_model.Reset, []credit_model.Grant, error) {
+	var rollovedGrants []credit_model.Grant
+
+	// Lock ledger for the subject
+	logger := c.logger.With("operation", "reset", "namespace", namespace, "subject", reset.Subject, "id", reset.ID)
+	lock, err := c.lockManager.Obtain(ctx, namespace, reset.Subject)
```

Comment:
Having an in memory lock will not scale if we are using multiple instances of the ledger for HA (or edge use-case).

Given that most of the code (besides Reset) manipulates a few rows, and given that this is using postgres I think it would be better to rely on postgres' built-in transaction support and [SELECT FOR UPDATE schemantics](https://www.postgresql.org/docs/current/sql-select.html#SQL-FOR-UPDATE-SHARE).

What you can do is to have a table for each ledger, e.g. 
```
CREATE TABLE ledgers (
   namespace VARCHAR(X)
   subject VARCHAR(X)
   -- cache fields such as credits, last reset record reference etc
)
-- + indexes
```

Then in case you want to forbid any updates to the specific ledger you can:
- start a transaction for each query
- do 
```
SELECT * FROM ledgers WHERE namespace = 'XXX'   and subject = 'XXX' FOR UPDATE;
```

This will make postgres sequence the calls instead of the in memory lock.

Generally speaking however this is still problematic as this will still be a global lock per namespace per subject.